### ISSUE TITLE
Puma 5.0.3 - issues with USR2/hot restart

### ISSUE THREAD

**Issue by majksner** [puma/puma#2454]
**Describe the bug**
I've updated to Puma 5.0.3 from 5.0.2 and as soon I've deployed the app to staging environment it failed immediately. 

**Reverting to 5.0.2 fixes the problem.**

**Puma config:**

```ruby
# frozen_string_literal: true

require('dotenv')

app_dir = File.expand_path('..', __dir__)

Dotenv.load("#{app_dir}/current/.env")

directory("#{app_dir}/current")
rackup("#{app_dir}/current/config.ru")
env = ENV.fetch('RAILS_ENV', 'development')
environment(env)

tag(env)

pidfile("#{app_dir}/shared/tmp/pids/puma.pid")
state_path("#{app_dir}/shared/tmp/pids/puma.state")
stdout_redirect("#{app_dir}/shared/log/puma_access.log", "#{app_dir}/shared/log/puma_error.log", true)

max_threads_count = ENV.fetch('RAILS_MAX_THREADS', 16)
threads(0, max_threads_count)

# port(ENV.fetch('PORT', 3000))

bind("unix:///#{app_dir}/shared/tmp/sockets/puma.sock")

workers(ENV.fetch('WEB_CONCURRENCY', 2))

preload_app!

prune_bundler

on_restart do
  ENV['BUNDLE_GEMFILE'] = ''
end

nakayoshi_fork

before_fork do
  require 'puma_worker_killer'
  # Default is every 6 hours
  PumaWorkerKiller.enable_rolling_restart
end

lowlevel_error_handler do |ex, environment|
  Raven.capture_exception(
    ex,
    message: ex.message,
    extra: { puma: environment },
    transaction: 'Puma'
  )
  # note the below is just a Rack response
  [
    500,
    {},
    [
      "An error has occurred, and engineers have been informed. Please
     reload the page. If you continue to have problems,
     contact support@company.com\n"
    ]
  ]
end
```

**Error message:**

```ruby
/var/www/app/shared/bundle/ruby/2.7.0/gems/railties-6.0.3.4/lib/rails/application.rb:362:in `initialize!': Application has been already initialized. (RuntimeError)
        from /var/www/app/releases/20201026134450/config/environment.rb:7:in `<main>'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/bootsnap-1.4.8/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23:in `require'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/bootsnap-1.4.8/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23:in `block in require_with_bootsnap_lfi'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/bootsnap-1.4.8/lib/bootsnap/load_path_cache/loaded_features_index.rb:92:in `register'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/bootsnap-1.4.8/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:22:in `require_with_bootsnap_lfi'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/bootsnap-1.4.8/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:31:in `require'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/zeitwerk-2.4.0/lib/zeitwerk/kernel.rb:34:in `require'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/bootsnap-1.4.8/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:53:in `require_relative'
        from /var/www/app/current/config.ru:5:in `block in <main>'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/rack-2.2.3/lib/rack/builder.rb:116:in `eval'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/rack-2.2.3/lib/rack/builder.rb:116:in `new_from_string'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/rack-2.2.3/lib/rack/builder.rb:105:in `load_file'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/rack-2.2.3/lib/rack/builder.rb:66:in `parse_file'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/puma-5.0.3/lib/puma/configuration.rb:331:in `load_rackup'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/puma-5.0.3/lib/puma/configuration.rb:257:in `app'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/puma-5.0.3/lib/puma/runner.rb:142:in `app'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/puma-5.0.3/lib/puma/runner.rb:146:in `start_server'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/puma-5.0.3/lib/puma/cluster/worker.rb:57:in `run'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/puma-5.0.3/lib/puma/cluster.rb:193:in `worker'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/puma-5.0.3/lib/puma/cluster.rb:96:in `block in spawn_worker'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/puma-5.0.3/lib/puma/cluster.rb:96:in `fork'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/puma-5.0.3/lib/puma/cluster.rb:96:in `spawn_worker'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/puma-5.0.3/lib/puma/cluster.rb:78:in `block in spawn_workers'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/puma-5.0.3/lib/puma/cluster.rb:71:in `times'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/puma-5.0.3/lib/puma/cluster.rb:71:in `spawn_workers'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/puma-5.0.3/lib/puma/cluster.rb:142:in `check_workers'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/puma-5.0.3/lib/puma/cluster.rb:408:in `run'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/puma-5.0.3/lib/puma/launcher.rb:171:in `run'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/puma-5.0.3/lib/puma/cli.rb:80:in `run'
        from /var/www/app/shared/bundle/ruby/2.7.0/gems/puma-5.0.3/bin/puma:10:in `<top (required)>'
        from /var/www/app/shared/bundle/ruby/2.7.0/bin/puma:23:in `load'
        from /var/www/app/shared/bundle/ruby/2.7.0/bin/puma:23:in `<main>'
        from /var/www/app/shared/bundle/ruby/2.7.0/bin/ruby_executable_hooks:24:in `eval'
        from /var/www/app/shared/bundle/ruby/2.7.0/bin/ruby_executable_hooks:24:in `<main>'
```

**To Reproduce**
Upgrade Puma from 5.0.2 to 5.0.3

**Expected behavior**
Rails app should start without errors.

**Desktop (please complete the following information):**
 - OS: Linux
 - Puma Version 5.0.3
 - Ruby 2.7.2
 - Rails Latest Version

**Comment by nateberkopec** [comment]
Please post your `application.rb`

**Comment by majksner** [comment]
Here you go:

```ruby
# frozen_string_literal: true

require_relative('boot')
require('rails')
require('active_record/railtie')
# require "active_storage/engine"
require('action_controller/railtie')
require('action_view/railtie')
require('action_mailer/railtie')
require('active_job/railtie')
require('action_cable/engine')
require('action_text/engine')
# require "action_mailbox/engine"
# require "rails/test_unit/railtie"
require('sprockets/railtie')

require_relative('../lib/app')

# Require the gems listed in Gemfile, including any gems
# you've limited to :test, :development, or :production.
Bundler.require(*Rails.groups)

module App
  class Application < Rails::Application
    config.load_defaults(6.0)

    config.time_zone = 'Europe/Brussels'

    # to be changed in future
    config.hosts = ['.company.com', '.company.test', 'localhost']

    config.ssl_options = { hsts: { preload: true } }
    config.force_ssl = true

    # config.session_store(:cookie_store, key: '_session', secure: true)
    config.session_store(:cache_store, key: '_session', secure: true)

    config.middleware.insert_before(0, Rack::Cors) do
      allow do
        origins 'company.test:3000', 'localhost:3000', %r{https*://.*?company\.com}
        resource '/jobs/*', headers: :any, methods: %i[get]
        resource '/partners/*', headers: :any, methods: %i[get]
        resource '/users/profiles*', headers: :any, methods: %i[get]
      end
    end

    config.active_job.queue_adapter = :sidekiq

    config.enable_dependency_loading = true

    config.reform.validations = :active_model

    config.middleware.use(Rack::Attack)

    config.exceptions_app = routes

    config.log_tags = %i[subdomain uuid]

    config.cache_store =
      :redis_cache_store,
      {
        db: ENV.fetch('REDIS_DB', 0),
        expires_in: 7.days,
        connect_timeout: 30,
        read_timeout: 5,
        write_timeout: 5,
        reconnect_attempts: 1,
        error_handler: lambda { |method:, returning:, exception:|
                         Raven.capture_exception(
                           exception,
                           level: 'warning',
                           tags: { method: method, returning: returning }
                         )
                       }
      }

    config.default_host = ENV.fetch('RAILS_HOST', 'app.company.com')

    config.action_mailer.asset_host = "https://#{config.default_host}"
    config.action_controller.asset_host = "https://#{config.default_host}"

    routes.default_url_options = { host: config.default_host, protocol: 'https' }
    config.action_mailer.default_url_options = { host: config.default_host }
    config.action_mailer.delivery_method = :smtp
    config.action_mailer.smtp_settings = {
      authentication: :plain,
      address: 'smtp.eu.mailgun.org',
      port: 587,
      domain: 'mg.company.com',
      user_name: 'postmaster@mg.company.com',
      password: Rails.application.credentials.mailgun[:password],
      enable_starttls_auto: true
    }

    config.action_cable.allowed_request_origins = ["https://#{config.default_host}"]
    config.action_cable.disable_request_forgery_protection = false
    config.action_cable.url = '/cable'


    Warning[:deprecated] = true

    config.generators do |g|
      g.view_specs(false)
      g.helper(false)
      g.assets(false)
      g.decorator(false)
      g.skip_routes(true)
    end
  end
end

```

**Comment by nateberkopec** [comment]
Does not reproduce on a blank `rails new` app (rails new, change Puma to 5.0.3, start)

Sorry for the comment about application.rb before, I was misreading your stacktrace.

**Comment by nateberkopec** [comment]
This looks like a production stacktrace, is that correct? How did you deploy this change?

**Comment by nateberkopec** [comment]
What does your `config/environment.rb` look like as well?

**Comment by nateberkopec** [comment]
I'm also interested if your app boots with 5.0.3 in your testing or development environments.

**Comment by majksner** [comment]
> What does your `config/environment.rb` look like as well?

Rails default:

```ruby
# frozen_string_literal: true

# Load the Rails application.
require_relative "application"

# Initialize the Rails application.
Rails.application.initialize!
```

**Comment by nateberkopec** [comment]
Ohhh thanks. Yours has the frozen string pragma and mine didn't, so there were different line numbers and I thought it was different. 

Still interested in your

1. Deployment strategy (hot/phased or other)
2. If the app boots locally w/5.0.3

**Comment by majksner** [comment]
> Ohhh thanks. Yours has the frozen string pragma and mine didn't, so there were different line numbers and I thought it was different.
> 
> Still interested in your
> 
>     1. Deployment strategy (hot/phased or other)
> 
>     2. If the app boots locally w/5.0.3

1. Depends on if there are migrations or not, I'm running `sudo systemctl reload puma.service` if there are migration, otherwise I run `sudo systemctl restart puma.service`

Here is my systemd unit:

```
[Unit]
Description=Puma Server (staging)
After=network.target

[Service]
Type=simple
User=deploy
Environment=RAILS_ENV=staging
EnvironmentFile=/etc/environment
WorkingDirectory=/var/www/app/current
ExecStart=/home/deploy/.rvm/bin/rvm default do bundle exec puma -C /var/www/app/shared/puma.rb
ExecStop=/home/deploy/.rvm/bin/rvm default do bundle exec pumactl -F /var/www/app/shared/puma.rb stop
ExecReload=/bin/kill -SIGUSR2 $MAINPID
TimeoutSec=15
Restart=always
KillMode=process

[Install]
WantedBy=multi-user.target
```

2. The app works in development and test mode

**Comment by majksner** [comment]
It works if I remove `preload_app!` from `puma.rb`

**Comment by nateberkopec** [comment]
@cjlarose hunches?

**Comment by nateberkopec** [comment]
@majksner I see. Your reload is a [USR2 hot restart](https://github.com/puma/puma/blob/master/docs/restart.md#hot-restart). `preload_app` is not and has never been supported for this configuration. 

Why it ever worked for you is up for debate 😆

**Comment by nateberkopec** [comment]
OK, still not reproing. I'm able to do a USR2 hot restart on a blank Rails app with `workers 2` and `preload_app!` locally.

**Comment by nateberkopec** [comment]
@majksner If you run the app locally, then `$ kill -USR2 <puma-master-process-pid>`, does it trigger the same error?

**Comment by majksner** [comment]
> OK, still not reproing. I'm able to do a USR2 hot restart on a blank Rails app with `workers 2` and `preload_app!` locally.

If I use `$ kill -USR2 $MAINPID` or `-USR1` with preload app on my staging environment it doesn't work still. So only removing `preload_app!` works.

> If you run the app locally, then $ kill -USR2 <puma-master-process-pid>, does it trigger the same error?

It doesn't

**Comment by nateberkopec** [comment]
Well, -USR1 I would expect not to work because preload_app is not supported for phased restart.

The fact that it doesn't work in your staging environment and works locally might be a clue...

**Comment by majksner** [comment]
It works when I put back `preload_app!` and remove `rackup("#{app_dir}/config.ru")`

**Comment by cjlarose** [comment]
Yeah, it looks like `preload_app!` was effectively broken in 5.0.3.

```ruby
# config.ru
STDERR.puts "Loading app"
run lambda { |env| [200, {"Content-Type" => "text/plain"}, ["Hello World"]] }
```

```sh
bundle exec ruby -Ilib bin/puma --workers 2 --prune-bundler --preload test/rackup/hello.ru
```

In 5.0.2, you can see that the application is loaded once

```
[33068] Puma starting in cluster mode...
[33068] * Version 5.0.2 (ruby 2.7.0-p0), codename: Spoony Bard
[33068] * Min threads: 0, max threads: 5
[33068] * Environment: development
[33068] * Process workers: 2
[33068] * Preloading application
Loading app
[33068] * Listening on http://0.0.0.0:9293
[33068] Use Ctrl-C to stop
[33068] - Worker 0 (pid: 33113) booted, phase: 0
[33068] - Worker 1 (pid: 33114) booted, phase: 0
```

In 5.0.3, the application is "preloaded" but the workers load the app again.

```
[33314] Puma starting in cluster mode...
[33314] * Version 5.0.3 (ruby 2.7.0-p0), codename: Spoony Bard
[33314] * Min threads: 0, max threads: 5
[33314] * Environment: development
[33314] * Process workers: 2
[33314] * Preloading application
Loading app
[33314] * Listening on http://0.0.0.0:9293
[33314] Use Ctrl-C to stop
Loading app
Loading app
[33314] - Worker 0 (pid: 33385) booted, phase: 0
[33314] - Worker 1 (pid: 33386) booted, phase: 0
```

I'll perform a `bisect` and dig into what broke it.

**Comment by cjlarose** [comment]
Well, as it turns out, I broke it in #2374. I think there's a missing History.md line for it, too.

I can open a fix.

**Comment by ianks** [comment]
Similar issues with us, except we are getting `Prometheus::Client::Registry::AlreadyRegisteredError`, meaning that metrics that were already registered during the Rack boot phase (from config.ru), were registered multiple times.

**Comment by nateberkopec** [comment]
Issue fixed and released as 5.0.4
